basic_pipeline/4_Caps.md

```diff
@@ -0,0 +1,116 @@
+Here is how you define a caps specification:
+1. First you need to specify the format module
+ ``` Elixir
+ defmodule Formats.Raw do
+   defstruct [:format, :framerate, :width, :height]
+ end
+ ```
+ Module name defines the type of the caps, however it is possible to pass some other options in a form of a struct. That is why we have defined a structure with the use of `defstruct`. Our format will be described with the following options:
+ + :format - pixel format, i.e. [I420](https://en.wikipedia.org/wiki/Chroma_subsampling) ([YUV](https://en.wikipedia.org/wiki/YUV)) or RGB888
+ + :framerate - number of frames per second, i.e. 30 (FPS)
+ + :width - width of the picture in pixels, i.e. 480 (px)
+ + :height - height of the picture in pixels, i.e. 300 (px)
+2. We specify the pad of the element with the format we have just defined, using the `:caps` option. For the purpose of an example, let it be the `:input` pad:
+ ```Elixir
+ def_input_pad(:input, 
+   demand_unit: :buffers, 
+   caps: [
+      {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 480, height: 300},
+      {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 720, height: 480}
+   ]
+ )
+ ```
+ As you can see, we pass a list of compatible formats, each described with the tuple, consisting of our module name, and the keywords list fulfilling the 
+ structure defined in that module. For the format's options, we can use the `range/2` or `one_of/1` specifier, which will modify the way in which the comparison between the caps specification and the actual caps received by the element is performed.
+
+3. Once the caps event will come to the element, the caps sent within that event will be compared to the caps specification we have just defined.
+ Caps sent in the event are confronted with each of the elements in the `caps:` list. If it matches even one of the caps formats in the list it means that caps are matching (alternative, OR).
+ To match the caps with the particular format (one from the list), the module (first element of the tuple in caps format description) and all the options must match (conjunction, AND). For each option, a value sent within the event is confronted with the specification of the option. The way comparison occurs is dependent on how we defined that option in the specification:
+ + We have used `framerate: range(30, 60)`, so will accept the framerate value in the given interval, between 30 and 60 FPS.
+ + We have also used `pixel_format: one_of([:I420, :I422]`, and that will accept caps, whose pixel format is either I420 or I422
+ + We have used a plain value to specify the `width` and the `height` of a picture - the caps will match if that option will be equal to the value passed in the specification 
```

Comment:
This description is a bit messy to me, can you try to reword it to make it more understandable?

Reply:
sure! I hope now it is more clear
done